More Jim Comments

jacobgardner opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/jacobgardner/monopoly-server/3)
<!-- Reviewable:end -->

jrhii committed:
dice object

added dice object

jacobgardner committed:
Squashed commit of the following:

commit 24815247a114af161b3b465d464b338ee0fd5a7a
Merge: b6dca0e 57bfbd5
Author: Jacob Gardner <jacob.v.gardner@gmail.com>
Date:   Thu Sep 22 17:29:36 2016 -0500

    more cleanup

commit b6dca0ee619bfd11745c51cf180dc10f149b0b9a
Author: Jacob Gardner <jacob.v.gardner@gmail.com>
Date:   Thu Sep 22 17:20:25 2016 -0500

    More pretty

commit c4004ea3d4c2cf182ac3be0217b3a231babf49c3
Author: Jacob Gardner <jacob.v.gardner@gmail.com>
Date:   Thu Sep 22 17:17:39 2016 -0500

    Make things pretty

commit 3474e7a18088cedb189e219d9d02ad0a78769a3c
Merge: 3e307b3 e70aef1
Author: Jacob Gardner <jacob.v.gardner@gmail.com>
Date:   Mon Sep 19 13:17:45 2016 -0500

    Merge pull request #1 from jacobgardner/master

    Approved!  For now.

jacobgardner commented on src/boardState.js:
```diff
@@ -1,7 +1,7 @@
 export default class BoardState{
```
Space after BoardState.  I forgot to do this.

jrhii replied:
done

jacobgardner commented on src/dice.js:
```diff
@@ -0,0 +1,20 @@
+import randomjs from 'random-js';
+
+export default class Dice{
```
Space after Dice... you get the idea.

jrhii replied:
done

jacobgardner commented on src/dice.js:
```diff
@@ -0,0 +1,20 @@
+    this.diceArray = this.random.dice(max, number);
+  }
+
+  isDoubles(){
```
Suggest changing this and below to getters.  We had words about this.

jrhii replied:
done

jacobgardner commented on src/landOnFunctions.js:
```diff
@@ -1,57 +1,67 @@
-export {standardProperty, railroad, utility, eventCard, noEvent, go, goToJail, incomeTax, luxuryTax};
+export {
```
Ok.  I was trying to figure out why this looked weird no matter what I did.  Then I realized that I never do this.   What you should do instead is delete this and add an export before each function. 

``` javascript
export function standardProperty(argsHere) {/* code here */}
export function utility(args...) { /* code */}
/* yadda yadda */
```

jrhii replied:
done

jacobgardner commented on src/landOnFunctions.js:
```diff
@@ -1,57 +1,67 @@
-  else{
-    emitter.emit(`promptPayment${activePlayer.socketID}`, this.ownerID, activePlayer, this.rent[houses]);
-  }
+    if(this.ownerID == null){
```
I know this works, but it feels almost sacrilege to use `this` in a normal function and not a method attached to a class.  What I'd suggest doing instead is passing in the property as the first argument and use that instead.  Alternatively you could make these methods apart of a property class and pass the property info into the constructor... I don't know I'd have to think about that way.

jacobgardner replied:
I guess the main reason I'm opposed to `this` in a normal function is because there is no class that it belongs to, `this` is ambiguous when looking at this code.  You would never open this file and think `this` is obviously a property object.

jrhii replied:
It's kind of weird since every other part of the class is laid out in the JSON file.  I have thought about doing what you said, then you have this class with 8 functions that don't apply to them.  Maybe I'm too hesitant to pass too many arguments in sometimes.

jacobgardner replied:
So you could have 8 classes.  

``` javascript
class BaseProperty {
 // All the shit they share
}

 class StandardProperty extends BaseProperty {
    landonFunction() {}
}

class Railroad extends BaseProperty {
   landonFunction() {}
}

etc.
```

jacobgardner replied:
As to passing in arguments you could use a PropertyFactory (a function/class that generates other objects)

``` javascript

import {StandardProperty, Railroad, Utilty, etc. } from './properties';

// This would replace that landon obj:
const PROPERTY_TYPES = {
     'standard': StandardProperty,
     'railroad': Railroad,
     'utility': Utility,
/// etc.
};

// This would be inside `./properties.js`
class BaseProperty {
    constructor(obj) {
        for (let key in object) {
            // All subclasses of BaseProperty will now have all the original 
            //    attributes from the json. 
            this[key] = obj[key];
        }
    }
};

function propertyFactory(objFromJson) {
        const Cls = PROPERTY_TYPES[objFromJson.propertyType];

        return new Cls(objFromJson);
}

```

jrhii replied:
later

jacobgardner replied:
I see you have made a confused emoji.  Anything I can help to clarify.

jacobgardner commented on src/landOnFunctions.js:
```diff
@@ -1,57 +1,67 @@
-    emitter.emit(`promptBuy${activePlayer.socketID}`, this.position);
-  }
+    if(this.ownerID == null){
+        emitter.emit(`promptBuy${activePlayer.socketID}`, this.position);
```
This line doesn't match the pattern set by the above 2...  Is this intentional?  I need to look deeper at the code for that.

jrhii replied:
Ah, I've found why the clients weren't reading utility names.  I had changed how I was going to do it and missed this one.

jrhii replied:
done

jacobgardner commented on src/landOnFunctions.js:
```diff
@@ -1,57 +1,67 @@
-    emitter.emit(`promptPayment${activePlayer.socketID}`, this.ownerID, activePlayer, this.rent[houses]);
-  }
+    if(this.ownerID == null){
+        emitter.emit(`promptBuy${activePlayer.socketID}`, this, activePlayer);
```
I don't like making part of the event name the player that you're wanting to prompt.  This will likely cause bugs down the line that will be harder to catch because things are tied to a very specific event name instead of a general event name covering all the players.  You're already passing in active player, inside the handler you should be able to get the player's socket you want to interact with there, right?

jrhii replied:
I'll have to look over where all I used it.  I think I started using it more because I am also trying to make it harder for other player to unintentionally or maliciously spam events and hit somebody else's.

Though, in this case the reason is because the listeners are turned on when a client connects with socket.io in main.js  As I understand it, each socket.on that opens is specific to that clients open socket  Since I am turning on listeners in the Monopoly.emitter object, it would otherwise turn on a listener each time a client connects and was causing issues.  Now, it turns on a client specific listener for each client.  Just as each client has their own specific socket listeners, they have their own monopoly listeners.  It also helps the server from taking in bad events emitted by clients as mentioned earlier.

jacobgardner replied:
I think my huge comment at the bottom avoids problems with this. As for malicious clients (i.e. Bobby) trying to force a promptWhatever to go through on not their turn, basically the solution is when a client replies, make sure that it's their turn before processing anything.

jacobgardner commented on src/monopoly.js:
```diff
@@ -1,189 +1,184 @@
+            'luxuryTax' : luxuryTax,
+        };
+
+        this.propertyArray = JSON.parse(fs.readFileSync('properties.JSON'));
```
JSON needs to be lowercase.  This doesn't work on linux.

jrhii replied:
done

jacobgardner commented on src/dice.js:
```diff
@@ -0,0 +1,20 @@
+  }
+
+  isDoubles(){
+    return this.diceArray.every(element => element == this.diceArray[0]);
```
`===`

jrhii replied:
done

jacobgardner commented on src/landOnFunctions.js:
```diff
@@ -1,57 +1,67 @@
-  else{
-    emitter.emit(`promptPayment${activePlayer.socketID}`, this.ownerID, activePlayer, this.rent[houses]);
-  }
+    if(this.ownerID == null){
```
Make sure you use `===` instead of `==`. `==` will choose the least common type of the two and compare that.  Basically, never use `==`.  `===` is more safe.

jrhii replied:
done

jacobgardner commented on src/landOnFunctions.js:
```diff
@@ -1,57 +1,67 @@
-  else{
-    emitter.emit(`promptPayment${activePlayer.socketID}`, this.ownerID, activePlayer, this.rent[houses]);
-  }
+    if(this.ownerID == null){
```
`===`

jrhii replied:
done

jacobgardner commented on src/landOnFunctions.js:
```diff
@@ -1,57 +1,67 @@
-  if(this.ownerID == null){
-    emitter.emit(`promptBuy${activePlayer.socketID}`, this.position);
-  }
+    if(this.ownerID == null){
```
`===`

jrhii replied:
done

jacobgardner commented on src/monopoly.js:
```diff
@@ -1,189 +1,184 @@
+        if(regArray.indexOf(regID) >= 0){
+            this.playerArray.push(new Player(name, regID, socketID));
+            console.log('Player ' + this.playerArray[this.playerArray.length - 1].person + ' added.');
+            if(regArray.length == this.playerArray.length){
```
`===`

jrhii replied:
done

jacobgardner commented on src/dice.js:
```diff
@@ -0,0 +1,20 @@
+import randomjs from 'random-js';
```
The spacing of this file should be made consistent with the rest of the project. 

To me, 2 spaces is hard to easily track which scope you're in when you're going 3 or more deep.

jrhii replied:
done

jacobgardner commented on src/main.js:
```diff
@@ -1,89 +1,89 @@
 const io = socketio(server);
 const monopolyGame = new Monopoly();
 
 io.on('connect', (socket) => {
```
Suggest moving `monopolyGame.emitter.on('promptWhatever',)` outside of `io.on('connect')` to eventually moved into another area.  Which I describe below...

So basically in the interest of supporting multiple monopoly games running simultaneously what you'll want is to have a startGame/joinGame event that the server listens to.  This module should probably eventually end up looking like the below probably.   Let me know if this makes sense.  It should fix some of the issues you've been hitting. 

``` javascript
const monopolyGamesByID = {}; // This should be globally accessible through the module.

function setUpEvents(game) {
     game.emitter.on('promptBuy', (property, player) => {
          player.emit('prompt', stuff)
     });

     game.emitter.on('start', () => {
          game.players.forEach((player) => {
               player.emit('GameStarted', game.boardstate, game.player, etc);
          });
     })

     // etc. etc.
}


io.on('connect', (socket) => {
     const state = {currentGame: null};

     socket.on('startGame', () => {
          const game = new Monopoly(...);
          state.currentGame = game;
          game.players.push(socketOrSomeObjectContainingSocket);
          const id = userGeneratedID || randomlyGeneratedID();
          monopolyGamesByID[id] = game;

          setUpEvents(game);

          socket.emit('gameCreated', id);
     });

     socket.on('joinGame', (gameID) => {
          const game = monopolyGamesByID[gameID];
          if (game && !game.inProgress) {
               game.players.push(socketOrSomeObjectContainingSocket);
               state.currentGame = game;
               if (game.seatsAvailable === 0) { 
                    game.start();
               } 
          }
     });

});
```

jrhii replied:
will evaluate this issue soon, involves my lengthy comment as well.

jrhii replied:
OK, looked at it and understand it.  this is my next thing im working on when it isn't 1AM

jrhii replied:
My solution seems messy, also will be changed somewhat changing landOnFunctions.js to omint instances of `this`

jacobgardner reviewed (commented):
Outside of those comments, you can probably also remove main2.js.  It's not needed any more.

jrhii committed:
Review Corrections, minor

fixed all the small stuff, working on big stuff next.

jrhii committed:
Review correction, less minor

moved events around as per suggestions

jacobgardner commented on src/landOnFunctions.js:
```diff
@@ -1,30 +1,119 @@
+            if(bool){
+                if(activePlayer.money < this.cost){
+                    io.to(activePlayer.socketID).emit('msg', `insufficient funds, purchase of ${this.nameStr} cancelled`);//IDEA: keep track of round info to send and make game logs
+                }
```
Typically, we put else on the same line as the closing brace it's elsing so it's clear which block it's the else of.

```
if (thing) {
    /// Stuff
} else {
    // Other Stuff
}
```

jacobgardner commented on src/landOnFunctions.js:
```diff
@@ -1,30 +1,119 @@
     if(this.ownerID === null){
-        emitter.emit(`promptBuy${activePlayer.socketID}`, this, activePlayer);
+        emitter.once(`confirmBuy`, (io, bool) => {
+            if(bool){
```
Keywords such as if/else/return, we typically make sure there's a space after them.  That way it's clear that we're using a keyword and not calling a function.

jacobgardner committed:
Stricter linting

jacobgardner committed:
Add linting manifesto

jacobgardner committed:
Some language fixes

jrhii committed:
Lint fixes

resuming work after week long down period.  This is 90% automatic fixing
by eslint

jrhii committed:
stuff

stuff

jrhii committed:
property object modifications

as per suggestion

jrhii committed:
Fixed Tax payments

Payments for income and luxury tax now work

jacobgardner commented on src/dummyClient/dummyClient2.js:
```diff
@@ -2,33 +2,33 @@ import fs from 'fs';
-//    fs.writeFileSync('clientBoardState.json', BoardState);
+    socket.on('promptRoll', (boardState) =>{
+        console.log('roll prompted');
+        fs.writeFileSync('clientBoardState_Tim.json', boardState);
```
Suggest renaming Tim to Harold.

jacobgardner commented on src/monopoly.js:
```diff
@@ -4,10 +4,10 @@ import Player from './player';
 import events from 'events';
 import randomjs from 'random-js';
-import {standardProperty, railroad, utility, eventCard, noEvent, go, goToJail, incomeTax, luxuryTax} from './landOnFunctions';
+import {StandardProperty, Railroad, Utility, EventCard, NoEvent, Go, GoToJail, IncomeTax, LuxuryTax} from './properties';
```
If you don't want to import all these, you could alternatively have './properties' export that PROPERTY_TYPES object.   Up to you.  It's just a matter of convenience.

jacobgardner commented on src/monopoly.js:
```diff
@@ -128,31 +125,35 @@ export default class Monopoly {
-            property.landOnFunction = PROPERTY_FUNCTIONS[property.functionID];
-        });
+        const JSONArray = JSON.parse(fs.readFileSync('properties.json'));
+        for (let property of JSONArray) {
```
:+1:

jacobgardner replied:
I think `property` can be `const` here.  I believe the way for loops work is that it's as if you declared the variable at the start of the inside of the block.  So you're not rebinding it.

jacobgardner commented on src/properties.js:
```diff
@@ -0,0 +1,145 @@
+class BaseProperty {
+    constructor(object) {
+        for (let key in object) {
+            this[key] = object[key];
```
Because you're just using an object parsed from JSON, this isn't an issue, but using `for (variable in obj)` can be somewhat dangerous in other situations.  Basically if your object inherits properties from another object, you'll iterate over all the properties of the parent object.  So if our object was a decedent of another we'd do:

```
for (const key in object) {
    if (object.hasOwnProperty(key)) {
        // Do stuff with key here. 
    }
}
```

Another example of this at the bottom of:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...in

Thinking about it now, we could do:

```
for (const key of Object.keys(object)) {
    // Do stuff with key
}
```

which is a lot sexier. 

OR..... if you're feeling adventurous, you could try modifying the Object 'class' so that it supports the `of` keyword so that your code looks like:

```
for (const [key, value] of object) {
/// We have both key and value :D
}
```

The implementation of that is left as an exercise to the reader.

All this is to say that this works perfectly fine and you don't need to change anything if you don't want to.

jacobgardner replied:
Minor thing.  You can make `key` `const` here.

jrhii committed:
Utilities Payment added

Added utilities payment, added monopoly boolean to properties.json.
Potential issue with RR rent but still need to add payment for multiple
railroads anyways.

jrhii committed:
minor issue

payment to bank works, payment to other players causing problem.

jrhii committed:
Minor Fix, Major step

Change in main fixes payment isssue.  Changed property name in player.js
for consistency sake.  Changed End of Turn event location for future
plans.

jrhii committed:
begun checking for monopolies

some issue wherein the check on line 151 always goes true.  everything
else should work

jrhii committed:
troubleshooting

jrhii committed:
troubleshoot

jrhii committed:
Jake found my problem

jrhii committed:
refined similar owner checks

jrhii committed:
Finished same owner (or monopoly) checks

fixed issue with array-sort having issues with sorting through nulls

jrhii committed:
Added files for event cards

jrhii committed:
More event card work

set up the load array function to also load event cards

jrhii committed:
Card funciton work

finished first card funciton

jrhii committed:
minor fix

wasn't included in last commit for whatever reason

jrhii committed:
eventCard payment function

Added function for standard payment event cards.  Corrected issues with
"advance to x" cards.

jrhii committed:
more event cards

minor update

jrhii committed:
more event cards

finished "Repairs" event, begun work on house building and selling.

jrhii committed:
buy and sell houses

Need to rework stuff based on rule revelations.  but can use a lot of
this still
